### PR TITLE
C wrapper for BCrypt

### PR DESCRIPTION
**Pull request by bre7**
### Checklist

- [ ] There are no breaking changes to public API.
    - [Change] `hash(_:cost:salt:)` returns a `String` instead of `Data`
- [ ] Circle CI is passing (code compiles and passes tests).
    - ⚠️ Fix https://github.com/vapor/vapor/blob/master/Sources/Development/main.swift#L29 (See api change mentioned above)
- [x] New test cases have been added where appropriate.
- [x] All new code has been commented with doc blocks `///`.

* OpenWall supports 2y while OpenBSD's doesn't (2y is replaced by 2b)
* OpenBSD is used in almost every wrapper (Python, Ruby, etc)

Included in this PR: OpenWall's (let me know if I should change it)

### CODE REVIEW

File: Sources/Random/URandom.swift [hunk 1]
```diff
@@ -46,4 +46,16 @@ public final class URandom: DataGenerator {
...
         return try read(numBytes: count)
     }
+
+    public func generateString(count: Int) throws -> String {
+        let bytes = try self.read(numBytes: count)
+
+        var string = ""
+        while (string.count < count) {
+            string += bytes
+                .compactMap({ String(bytes: [$0], encoding: .utf8) })
```
Comment (bre7) [inline Sources/Random/URandom.swift]: ⚠️ Swift 4.1 +

File: Sources/Crypto/BCrypt/BCryptDigest.swift [hunk 2]
```diff
@@ -1,60 +1,108 @@
...
+    public func hash(_ message: String, cost: UInt = 12) throws -> String {
+        let salt = try generateSalt(cost: cost)
+        return try hash(message, salt: salt)
+    }
+
+    public func hash(_ message: String, cost: UInt = 12, salt: String) throws -> String {
+        var pointer: UnsafeMutableRawPointer? = UnsafeMutableRawPointer.allocate(byteCount: 40 * MemoryLayout<UInt8>.stride, alignment: MemoryLayout<UInt8>.alignment)
+        var dstPointerSize = Int32(40)
+
+        guard let encryptedRaw = crypt_ra(
```
Comment (bre7) [inline Sources/Crypto/BCrypt/BCryptDigest.swift]: https://github.com/bre7/crypto/blob/0757facda5cb22cb48b7e004f5ca94355042e948/Sources/libbcrypt/wrapper.c#L171

Comment (bre7) [reply Sources/Crypto/BCrypt/BCryptDigest.swift]: https://github.com/bre7/crypto/blob/0757facda5cb22cb48b7e004f5ca94355042e948/Sources/libbcrypt/crypt_blowfish.c#L816

File: Package.swift [hunk 3]
```diff
@@ -4,7 +4,7 @@ import PackageDescription
 let package = Package(
     name: "Crypto",
     products: [
-        .library(name: "Crypto", targets: ["Crypto"]),
+        .library(name: "Crypto", targets: ["Crypto", "libbcrypt"]),
```
Comment (tanner0101) [inline Package.swift]: I think this change can be removed, we don't want to publicize `libbcrypt`. That's an implementation detail.

File: Package.swift [hunk 4]
```diff
@@ -15,7 +15,8 @@ let package = Package(
         .package(url: "https://github.com/apple/swift-nio-ssl.git", from: "1.0.0"),
     ],
     targets: [
-        .target(name: "Crypto", dependencies: ["Async", "Bits", "Core", "COperatingSystem", "Debugging", "NIOOpenSSL", "Random"]),
+    	.target(name: "libbcrypt"),
```
Comment (tanner0101) [inline Package.swift]: spacing, and targets should be in alphabetical order.

File: Sources/Crypto/BCrypt/BCryptDigest.swift [hunk 5]
```diff
@@ -1,60 +1,108 @@
...
+            random,
+            Int32(random.count) //Int32(entropy.utf8.count / MemoryLayout<UInt8>.size)
+        ) else {
+            throw CryptoError(identifier: "unableToGenerateSalt", reason: "Unable to generate BCrypt salt")
+        }
+
+        return String(cString: saltRaw)
+    }
+
+    public func hash(_ message: String, cost: UInt = 12) throws -> String {
```
Comment (tanner0101) [inline Sources/Crypto/BCrypt/BCryptDigest.swift]: All public methods must have doc blocks that explain all parameters and have code examples.

Comment (bre7) [reply Sources/Crypto/BCrypt/BCryptDigest.swift]: Yeah, ignore this PR. Submitting the OpenBSD version

**Comment by bre7** [comment]
See #55

**Pull request closed without merging**